scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +732,214 @@ class ProcessMonitor(ScalyrMonitor):
+
+        running_pids_set = set(running_pids)
+
+        for pid, process_metrics in self.metrics_history.items():
+            for _metric, _metric_values in process_metrics.items():
+                if not self.running_total_metrics.get(_metric):
+                    self.running_total_metrics[_metric] = 0
+                if _metric.is_cumulative:
+                    if pid in running_pids_set:
+                        if len(_metric_values) <= 1:
+                            self.running_total_metrics[_metric] += (_metric_values[-1] if _metric_values else 0)
+                        else:
+                            self.running_total_metrics[_metric] += (_metric_values[-1] - _metric_values[-2])
+                else:
+                    # absolute metric - accumulate the last reported value
+                    self.running_total_metrics[_metric] += _metric_values[-1]
+
+    def _remove_dead_processes(self, running_pids):
+        # once the _calculate_running_total has calculated the running total of the
+        # metrics for the current epoch, remove the entries for the process id in the history
+        # NOTE: the removal of the contributions (if applicable) should already be done so
+        # removing the entry from the history is safe.
+
+        all_pids = self.metrics_history.keys()
+        for _pid_to_remove in list(set(all_pids) - set(running_pids)):
+            # for all the absolute metrics, decrease the count that the dead processes accounted for
+            del self.metrics_history[_pid_to_remove]
+
+        # if no processes are running, there is no reason to report the running metric data
+        if not running_pids:
+            self.running_total_metrics = {}
 
     def gather_sample(self):
-        """Collect the per-process metrics for the monitored process.
+        """Collect the per-process tracker for the monitored process(es).
 
-        If the process is no longer running, then attempts to match a new one.
+        For multiple processes, there are a few cases we need to poll for:
+
+        For `n` processes, if there are < n running processes at any given point,
+        we should see if there are any pids or matching expression that gives pids
+        that is currently not running, if so, get its tracker and run it.
+        It is possible that one or more processes ran its course, but we can't assume that
+        and we should keep polling for it.
         """
-        if self.__pid is not None and not self.__is_running():
-            self.__set_process(None)
 
-        if self.__pid is None:
-            self.__set_process(self.__select_process())
+        trackers = []
+        for _pid in list(self.__select_processes()):
+            trackers.append(ProcessTracker(_pid, self._logger, self.__id))
+
+        self._reset_absolute_metrics()
 
-        for gather in self.__gathers:
-            gather.run_single_cycle()
+        for _tracker in trackers:
+            _tracker.set_gathers()
```

Comment:
Can we just have the gathers set at `ProcessTracker` initialization?  It seems a little odd to have to call `set_gathers` to use it.